presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java

```diff
@@ -106,6 +127,48 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, Co
         }
     }
 
+    private boolean isTimestampUpperBoundForcePushdownEnabled(ConnectorSession session, String topic)
```

Comment:
> Please also confirm that we are certain that if topic is in `LOG_APPEND` mode, we are sure that timestamps are alway recored in non-decreasing order when looking at messages by offset.
> I would expect so, but I do not know details of Kafka, so maybe if there is time skew between machines participating in the communication this does not need to always be true.
> If above does not hold we cannot filter agains upper bound at all.

In `LogAppendTime ` mode, the timestamp is a monotonically increasing function of the offset.  Note that this is in the context of a **single partition**.

> As for the `CREATE_TIME` mode. I am not sure that we want to expose the config property which enables upper bound filtering in that mode.
> It is easy to shoot oneself in the foot with the property, if user does not understand it's data, and filtering logic very well.
> Given fact that upper bound filtering is not that useful IMO (I would expect that you mostly need filtering to incrementally process new batches of data which are added to the topic; and for that lower bound filtering is enough), I would disable upper bound filter for `CREATE_TIME` mode.

That was my initial reaction, as well, but the reason the toggle was added is because one might have an application that is guaranteed to publish messages in the order they were created and users might want to take advantage of that.  By default, upper bound filtering is disabled in `CreateTime` mode.

Reply:
I still think it is a bad idea to use an upper bound filter in the `CreateTime`mode as there is absolutely no guarantee that the header timestamp created on the client side will be in sync with the timestamp used to create the timestamp-to-offset index on the server side. But in the query push-down that index is used to translate to an offset whereas in the result the value of the timestamp in the record is shown, which might be the same, but could also diverge quite a bit. So that might be very confusing for a user and lead to situations i have shown before, even if programmer does not overwrite the timestamp header and just uses the default behavior.

Reply:
@gschmutz Doesn't `KafkaConsumer.offsetsForTimes` work in a way that it resolves the timestamp to offset mapping on the server side. And resolution is based on exactly the one timestamp field on messages which is set either on log appending (in LOG_APPEND mode) or on message creation (in CREATE_MODE).
Are there two different timestamp fields, one reported in the message, and other one used for resolution? I don't think there are.